Sources/FoundationEssentials/String/StringProtocol+Essentials.swift

```diff
@@ -149,54 +149,96 @@ extension StringProtocol {
-
-        // It's possible that passed-in indices are not on unicode scalar boundaries, such as when they're UTF-16 indices.
-        // Expand the bounds to ensure they are so we can meaningfully iterate their UTF8 views.
-        let r = unicodeScalars._boundaryAlignedRange(aRange)
-        let result = utf8._getBlock(for: [.findStart, .findEnd, .stopAtLineSeparators], in: r)
-
-        guard let start = result.start else {
-            guard let end = result.end else {
-                return startIndex ..< endIndex
-            }
-            return startIndex ..< end
-        }
-
-        guard let upper = result.end else {
-            return start ..< endIndex
+extension StringProtocol {
+    @inline(never)
+    internal func _lineBounds(
+        around range: some RangeExpression<Index>
+    ) -> (start: Index, end: Index, contentsEnd: Index) {
+        // Avoid generic paths in the common case by manually specializing on `String` and
+        // `Substring`. Note that we're only ever calling `_lineBounds` on a `Substring`; this is
+        // to reduce the code size overhead of having to specialize it multiple times (at a slight
+        // cost to runtime performance).
+        if let s = _specializingCast(self, to: String.self) {
+            let range = s.unicodeScalars._boundaryAlignedRange(range)
+            return s[...].utf8._lineBounds(around: range)
+        } else if let s = _specializingCast(self, to: Substring.self) {
+            let range = s.unicodeScalars._boundaryAlignedRange(range)
+            return s.utf8._lineBounds(around: range)
+        } else {
+            // Unexpected case. `StringProtocol`'s UTF-8 view is not properly constrained, so we
+            // need to convert `self` to a Substring and carefully convert indices between the two
+            // collections before & after the _lineBounds call.
+            let range = self.unicodeScalars._boundaryAlignedRange(range)
+
+            let startUTF8Offset = self.utf8.distance(from: self.startIndex, to: range.lowerBound)
+            let utf8Count = self.utf8.distance(from: range.lowerBound, to: range.upperBound)
+
+            let s = Substring(self)
+            let start = s.utf8.index(s.startIndex, offsetBy: startUTF8Offset)
+            let end = s.utf8.index(start, offsetBy: utf8Count)
+            let r = s.utf8._lineBounds(around: start ..< end)
+
+            let resultUTF8Offsets = (
+                start: s.utf8.distance(from: s.startIndex, to: r.start),
+                end: s.utf8.distance(from: s.startIndex, to: r.end),
+                contentsEnd: s.utf8.distance(from: s.startIndex, to: r.contentsEnd))
+            return (
+                start: self.utf8.index(self.startIndex, offsetBy: resultUTF8Offsets.start),
+                end: self.utf8.index(self.startIndex, offsetBy: resultUTF8Offsets.end),
+                contentsEnd: self.utf8.index(self.startIndex, offsetBy: resultUTF8Offsets.contentsEnd))
```

Comment:
Following the above question -- Do we still need this if we already sanitize the indices before passing them to `_lineBounds(around:)`? Looking at the underlying `_getBlock`'s implementation, I think the returned indices should already be properly aligned if the indices passed to are already utf8-aligned. 

If it's still needed, do you think it would be reasonable to sink this index adjustment into `_lineBounds`'s implementation? The idea is that if the client uses `_lineBounds` on a utf8 view, they can be assured that the returned indices would be utf8-aligned. This way we can consolidate this and that in `_paragraphBounds(:)`

Reply:
Yes! These lines aren't for aligning the indices; rather, they are finding the positions in one collection that correspond to the positions we constructed in another.

(Indices are like pointers -- they are inherently tied to the particular collection instance that vended them, and aren't generally valid in any other. `String.Index` is mostly just an offset into UTF-8 (or UTF-16) storage; however, this offset value can be different between `self` and `s`, rendering index values incompatible between them. (For example, `self` could be using UTF-16 indices while `s` is UTF-8, or `self` can be a substring-like `StringProtocol` type where the `startIndex` isn't at offset zero. On this code path we already know the client is doing something weird, because we encountered a `StringProtocol` type that isn't `String` nor `Substring` -- we cannot safely assume anything about the internal structure of it.)

Reply:
Ah I see. This is to convert the indices of the `Substring` we constructed back to those of `self`. It has nothing to do with alignment.